File: src/main/java/com/milaboratory/mixcr/trees/ClonesRebase.java
package com.milaboratory.mixcr.trees;

import com.milaboratory.core.Range;
import com.milaboratory.core.alignment.Aligner;
import com.milaboratory.core.alignment.AlignmentScoring;
import com.milaboratory.core.mutations.Mutations;
import com.milaboratory.core.sequence.NucleotideSequence;
import com.milaboratory.core.sequence.SequenceBuilder;
import com.milaboratory.mixcr.util.RangeInfo;

class ClonesRebase {
    private final NucleotideSequence VSequence1;
    private final AlignmentScoring<NucleotideSequence> VScoring;
    private final AlignmentScoring<NucleotideSequence> NDNScoring;
    private final NucleotideSequence JSequence1;
    private final AlignmentScoring<NucleotideSequence> JScoring;

    ClonesRebase(
            NucleotideSequence VSequence1,
            AlignmentScoring<NucleotideSequence> VScoring,
            AlignmentScoring<NucleotideSequence> NDNScoring,
            NucleotideSequence JSequence1,
            AlignmentScoring<NucleotideSequence> JScoring
    ) {
        this.VSequence1 = VSequence1;
        this.VScoring = VScoring;
        this.JSequence1 = JSequence1;
        this.NDNScoring = NDNScoring;
        this.JScoring = JScoring;
    }

    CloneWithMutationsFromReconstructedRoot rebaseClone(RootInfo rootInfo, MutationsFromVJGermline mutationsFromVJGermline, CloneWrapper cloneWrapper) {
        Range NDNRangeInKnownNDN = NDNRangeInKnownNDN(mutationsFromVJGermline, rootInfo.getVRangeInCDR3(), rootInfo.getJRangeInCDR3());

        MutationsWithRange VMutationsInCDR3WithoutNDN = mutationsFromVJGermline.getVMutationsInCDR3WithoutNDN();
        Range wasVRangeInCDR3 = mutationsFromVJGermline.getVMutationsInCDR3WithoutNDN().getRangeInfo().getRange();
        Range VRange = new Range(wasVRangeInCDR3.getLower(), rootInfo.getVRangeInCDR3().getUpper());
        //can skip empty VRange because we will not include first mutations (empty range always will mutate to empty range)
        if (!VRange.isEmpty()) {
            Range VMutationsWithinNDNRange = mutationsFromVJGermline.getKnownVMutationsWithinNDN().getSecond()
                    .intersection(VRange);
            VMutationsWithinNDNRange = VMutationsWithinNDNRange != null ? VMutationsWithinNDNRange
                    : new Range(wasVRangeInCDR3.getUpper(), wasVRangeInCDR3.getUpper());
            int lengthDelta = 0;
            if (!VMutationsWithinNDNRange.isEmpty()) {
                MutationsWithRange VMutationsToAdd = new MutationsWithRange(
                        VSequence1,
                        mutationsFromVJGermline.getKnownVMutationsWithinNDN().getFirst(),
                        new RangeInfo(VMutationsWithinNDNRange, false)
                );
                VMutationsInCDR3WithoutNDN = VMutationsInCDR3WithoutNDN.combineWithTheSameMutationsRight(VMutationsToAdd);
                lengthDelta += VMutationsToAdd.lengthDelta();
            }
            Range rangeToAlign = new Range(VMutationsWithinNDNRange.getUpper(), rootInfo.getVRangeInCDR3().getUpper());
            if (!rangeToAlign.isEmpty() && !rangeToAlign.isReverse()) {
                Mutations<NucleotideSequence> absoluteMutations = Aligner.alignGlobal(
                        VScoring,
                        VSequence1,
                        mutationsFromVJGermline.getKnownNDN(),
                        rangeToAlign.getLower(),
                        rangeToAlign.length(),
                        VMutationsWithinNDNRange.length() + lengthDelta,
                        rangeToAlign.length()
                ).getAbsoluteMutations();
                VMutationsInCDR3WithoutNDN = VMutationsInCDR3WithoutNDN.combineWithMutationsToTheRight(absoluteMutations, rangeToAlign);
            }
            NDNRangeInKnownNDN = new Range(NDNRangeInKnownNDN.getLower() + lengthDelta, NDNRangeInKnownNDN.getUpper());
        }

        MutationsWithRange JMutationsInCDR3WithoutNDN = mutationsFromVJGermline.getJMutationsInCDR3WithoutNDN();
        Range wasJRangeInCDR3 = mutationsFromVJGermline.getJMutationsInCDR3WithoutNDN().getRangeInfo().getRange();
        Range JRange = new Range(rootInfo.getJRangeInCDR3().getLower(), wasJRangeInCDR3.getLower());
        Range JMutationsWithinNDNRange = mutationsFromVJGermline.getKnownJMutationsWithinNDN().getSecond()
                .intersection(JRange);
        JMutationsWithinNDNRange = JMutationsWithinNDNRange != null ? JMutationsWithinNDNRange
                : new Range(wasJRangeInCDR3.getLower(), wasJRangeInCDR3.getLower());
        int lengthDelta = 0;
        MutationsWithRange JMutationsToAdd = new MutationsWithRange(
                JSequence1,
                mutationsFromVJGermline.getKnownJMutationsWithinNDN().getFirst(),
                new RangeInfo(JMutationsWithinNDNRange, true)
        );
        JMutationsInCDR3WithoutNDN = JMutationsInCDR3WithoutNDN.combineWithTheSameMutationsLeft(JMutationsToAdd);
        lengthDelta += JMutationsToAdd.lengthDelta();
        Range rangeToAlign = new Range(rootInfo.getJRangeInCDR3().getLower(), JMutationsWithinNDNRange.getLower());
        if (!rangeToAlign.isEmpty() && !rangeToAlign.isReverse()) {
            Mutations<NucleotideSequence> absoluteMutations = Aligner.alignGlobal(
                    JScoring,
                    JSequence1,
                    mutationsFromVJGermline.getKnownNDN(),
                    rangeToAlign.getLower(),
                    rangeToAlign.length(),
                    NDNRangeInKnownNDN.getUpper() - lengthDelta,
                    rangeToAlign.length()
            ).getAbsoluteMutations();
            JMutationsInCDR3WithoutNDN = JMutationsInCDR3WithoutNDN.combineWithMutationsToTheLeft(absoluteMutations, rangeToAlign);
        }
        NDNRangeInKnownNDN = new Range(NDNRangeInKnownNDN.getLower(), NDNRangeInKnownNDN.getUpper() - lengthDelta);
        MutationsDescription mutations = new MutationsDescription(
                mutationsFromVJGermline.getVMutationsWithoutCDR3(),
                VMutationsInCDR3WithoutNDN,
                NDNMutations(rootInfo.getReconstructedNDN(), mutationsFromVJGermline.getKnownNDN().getRange(NDNRangeInKnownNDN)),
                JMutationsInCDR3WithoutNDN,
                mutationsFromVJGermline.getJMutationsWithoutCDR3()
        );
        return new CloneWithMutationsFromReconstructedRoot(mutations, mutationsFromVJGermline, cloneWrapper);
    }

    MutationsDescription rebaseMutations(
            MutationsDescription originalNode,
            RootInfo originalRoot,
            RootInfo rebaseTo
    ) {
        NucleotideSequence originalKnownNDN = originalNode.getKnownNDN().buildSequence();
        MutationsWithRange VMutationsInCDR3WithoutNDN;
        if (originalRoot.getVRangeInCDR3().length() < rebaseTo.getVRangeInCDR3().length()) {
            Range difference = new Range(originalRoot.getVRangeInCDR3().getUpper(), rebaseTo.getVRangeInCDR3().getUpper());
            VMutationsInCDR3WithoutNDN = originalNode.getVMutationsInCDR3WithoutNDN()
                    .combineWithMutationsToTheRight(
                            Aligner.alignGlobal(
                                    VScoring,
                                    VSequence1,
                                    originalKnownNDN,
                                    difference.getLower(),
                                    difference.length(),
                                    0,
                                    difference.length()
                            ).getAbsoluteMutations(),
                            difference
                    );
        } else if (rebaseTo.getVRangeInCDR3().length() < originalRoot.getVRangeInCDR3().length()) {
            VMutationsInCDR3WithoutNDN = new MutationsWithRange(
                    VSequence1,
                    originalNode.getVMutationsInCDR3WithoutNDN().getMutations(),
                    new RangeInfo(
                            rebaseTo.getVRangeInCDR3(),
                            false
                    )
            );
        } else {
            VMutationsInCDR3WithoutNDN = originalNode.getVMutationsInCDR3WithoutNDN();
        }

        SequenceBuilder<NucleotideSequence> knownNDNBuilder = NucleotideSequence.ALPHABET.createBuilder();

        if (rebaseTo.getVRangeInCDR3().length() < originalRoot.getVRangeInCDR3().length()) {
            knownNDNBuilder.append(MutationsUtils.buildSequence(
                    VSequence1,
                    originalNode.getVMutationsInCDR3WithoutNDN().getMutations(),
                    new RangeInfo(
                            new Range(rebaseTo.getVRangeInCDR3().getUpper(), originalRoot.getVRangeInCDR3().getUpper()),
                            false
                    )
            ));
        }

        knownNDNBuilder.append(originalKnownNDN.getRange(
                Math.max(0, rebaseTo.getVRangeInCDR3().length() - originalRoot.getVRangeInCDR3().length()),
                originalKnownNDN.size() - Math.max(0, rebaseTo.getJRangeInCDR3().length() - originalRoot.getJRangeInCDR3().length())
        ));

        if (rebaseTo.getJRangeInCDR3().length() < originalRoot.getJRangeInCDR3().length()) {
            knownNDNBuilder.append(MutationsUtils.buildSequence(
                    JSequence1,
                    originalNode.getJMutationsInCDR3WithoutNDN().getMutations(),
                    new RangeInfo(
                            new Range(originalRoot.getJRangeInCDR3().getLower(), rebaseTo.getJRangeInCDR3().getLower()),
                            false
                    )
            ));
        }

        NucleotideSequence rebasedKnownNDN = knownNDNBuilder.createAndDestroy();

        MutationsWithRange JMutationsInCDR3WithoutNDN;
        if (originalRoot.getJRangeInCDR3().length() < rebaseTo.getJRangeInCDR3().length()) {
            Range difference = new Range(rebaseTo.getJRangeInCDR3().getLower(), originalRoot.getJRangeInCDR3().getLower());
            JMutationsInCDR3WithoutNDN = originalNode.getJMutationsInCDR3WithoutNDN()
                    .combineWithMutationsToTheLeft(
                            Aligner.alignGlobal(
                                    JScoring,
                                    JSequence1,
                                    originalKnownNDN,
                                    difference.getLower(),
                                    difference.length(),
                                    originalKnownNDN.size() - difference.length(),
                                    difference.length()
                            ).getAbsoluteMutations(),
                            difference
                    );
        } else if (rebaseTo.getJRangeInCDR3().length() < originalRoot.getJRangeInCDR3().length()) {
            JMutationsInCDR3WithoutNDN = new MutationsWithRange(
                    JSequence1,
                    originalNode.getJMutationsInCDR3WithoutNDN().getMutations(),
                    new RangeInfo(
                            rebaseTo.getJRangeInCDR3(),
                            false
                    )
            );
        } else {
            JMutationsInCDR3WithoutNDN = originalNode.getJMutationsInCDR3WithoutNDN();
        }


        return new MutationsDescription(
                originalNode.getVMutationsWithoutCDR3(),
                VMutationsInCDR3WithoutNDN,
                NDNMutations(rebaseTo.getReconstructedNDN(), rebasedKnownNDN),
                JMutationsInCDR3WithoutNDN,
                originalNode.getJMutationsWithoutCDR3()
        );
    }


    static Range NDNRangeInKnownNDN(MutationsFromVJGermline mutations, Range VRangeInCDR3, Range JRangeInCDR3) {
        return new Range(
                VRangeInCDR3.length() - mutations.getVMutationsInCDR3WithoutNDN().getRangeInfo().getRange().length(),
                mutations.getKnownNDN().size() - (JRangeInCDR3.length() - mutations.getJMutationsInCDR3WithoutNDN().getRangeInfo().getRange().length())
        );
    }

    MutationsWithRange NDNMutations(NucleotideSequence first, NucleotideSequence second) {
        return new MutationsWithRange(
                first,
                Aligner.alignGlobal(
                        NDNScoring,
                        first,
                        second
                ).getAbsoluteMutations(),
                new RangeInfo(new Range(0, first.size()), true)
        );
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/trees/ClonesRebase.kt
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
@file:Suppress("FunctionName", "LocalVariableName")

package com.milaboratory.mixcr.trees

import com.milaboratory.core.Range
import com.milaboratory.core.alignment.Aligner
import com.milaboratory.core.sequence.NucleotideSequence
import com.milaboratory.mixcr.util.VJPair
import com.milaboratory.mixcr.util.plus

class ClonesRebase(
    private val sequence1: VJPair<NucleotideSequence>,
    private val scoringSet: ScoringSet
) {
    fun rebaseClone(
        rootInfo: RootInfo,
        mutationsFromVJGermline: MutationsFromVJGermline,
        cloneWrapper: CloneWrapper
    ): CloneWithMutationsFromReconstructedRoot {
        val VMutationsInCDR3WithoutNDN = Aligner.alignGlobal(
            scoringSet.V.scoring,
            sequence1.V,
            mutationsFromVJGermline.CDR3,
            rootInfo.rangeInCDR3.V.lower,
            rootInfo.rangeInCDR3.V.length(),
            0,
            rootInfo.rangeInCDR3.V.length()
        ).absoluteMutations

        val JMutationsInCDR3WithoutNDN = Aligner.alignGlobal(
            scoringSet.J.scoring,
            sequence1.J,
            mutationsFromVJGermline.CDR3,
            rootInfo.rangeInCDR3.J.lower,
            rootInfo.rangeInCDR3.J.length(),
            mutationsFromVJGermline.CDR3.size() - rootInfo.rangeInCDR3.J.length(),
            rootInfo.rangeInCDR3.J.length()
        ).absoluteMutations

        val NDNRangeToCompare = Range(
            rootInfo.rangeInCDR3.V.length() + VMutationsInCDR3WithoutNDN.lengthDelta,
            mutationsFromVJGermline.CDR3.size() - rootInfo.rangeInCDR3.J.length() - JMutationsInCDR3WithoutNDN.lengthDelta
        )
        check(!NDNRangeToCompare.isReverse) {
            "V CDR3 part and J CDR3 part intersect in ${cloneWrapper.VJBase}"
        }
        val compareWithNDN = when {
            NDNRangeToCompare.isEmpty -> NucleotideSequence.EMPTY
            else -> mutationsFromVJGermline.CDR3.getRange(NDNRangeToCompare)
        }
        val NDNMutations = Aligner.alignGlobal(
            scoringSet.NDN.scoring,
            rootInfo.reconstructedNDN,
            compareWithNDN
        ).absoluteMutations
        val result = CloneWithMutationsFromReconstructedRoot(
            MutationsSet(
                VGeneMutations(
                    mutationsFromVJGermline.mutations.V,
                    PartInCDR3(rootInfo.rangeInCDR3.V, VMutationsInCDR3WithoutNDN)
                ),
                NDNMutations(NDNMutations),
                JGeneMutations(
                    PartInCDR3(rootInfo.rangeInCDR3.J, JMutationsInCDR3WithoutNDN),
                    mutationsFromVJGermline.mutations.J
                )
            ),
            mutationsFromVJGermline,
            cloneWrapper
        )
        check(result.mutationsFromVJGermline.CDR3 == result.mutationsSet.buildCDR3(rootInfo)) {
            "Rebased CDR3 not equal to original in ${cloneWrapper.VJBase}"
        }
        return result
    }

    fun rebaseMutations(
        originalNode: MutationsSet,
        originalRoot: RootInfo,
        rebaseTo: RootInfo
    ): MutationsSet {
        val CDR3 = originalNode.buildCDR3(originalRoot)

        val VMutationsInCDR3WithoutNDN = Aligner.alignGlobal(
            scoringSet.V.scoring,
            sequence1.V,
            CDR3,
            rebaseTo.rangeInCDR3.V.lower,
            rebaseTo.rangeInCDR3.V.length(),
            0,
            rebaseTo.rangeInCDR3.V.length()
        ).absoluteMutations
        val JMutationsInCDR3WithoutNDN = Aligner.alignGlobal(
            scoringSet.J.scoring,
            sequence1.J,
            CDR3,
            rebaseTo.rangeInCDR3.J.lower,
            rebaseTo.rangeInCDR3.J.length(),
            CDR3.size() - rebaseTo.rangeInCDR3.J.length(),
            rebaseTo.rangeInCDR3.J.length()
        ).absoluteMutations

        val NDNRangeToCompare = Range(
            rebaseTo.rangeInCDR3.V.length() + VMutationsInCDR3WithoutNDN.lengthDelta,
            CDR3.size() - rebaseTo.rangeInCDR3.J.length() - JMutationsInCDR3WithoutNDN.lengthDelta
        )
        check(!NDNRangeToCompare.isReverse) {
            "V CDR3 part and J CDR3 part intersect in ${originalRoot.VJBase}"
        }
        val compareWithNDN = when {
            NDNRangeToCompare.isEmpty -> NucleotideSequence.EMPTY
            else -> CDR3.getRange(NDNRangeToCompare)
        }
        val NDNMutations = Aligner.alignGlobal(
            scoringSet.NDN.scoring,
            rebaseTo.reconstructedNDN,
            compareWithNDN
        ).absoluteMutations

        val result = MutationsSet(
            VGeneMutations(
                mutations = originalNode.mutations.V.mutationsOutsideOfCDR3,
                partInCDR3 = PartInCDR3(
                    rebaseTo.rangeInCDR3.V,
                    VMutationsInCDR3WithoutNDN
                )
            ),
            NDNMutations(NDNMutations),
            JGeneMutations(
                mutations = originalNode.mutations.J.mutationsOutsideOfCDR3,
                partInCDR3 = PartInCDR3(
                    rebaseTo.rangeInCDR3.J,
                    JMutationsInCDR3WithoutNDN
                )
            )
        )
        check(originalNode.buildCDR3(originalRoot) == result.buildCDR3(rebaseTo)) {
            "Rebased CDR3 not equal to original"
        }
        return result
    }

    private fun MutationsSet.buildCDR3(rootInfo: RootInfo) = MutationsUtils.buildSequence(
        rootInfo.sequence1.V,
        mutations.V.partInCDR3.mutations,
        rootInfo.rangeInCDR3.V
    ) + NDNMutations.mutations.mutate(rootInfo.reconstructedNDN) +
            MutationsUtils.buildSequence(
                rootInfo.sequence1.J,
                mutations.J.partInCDR3.mutations,
                rootInfo.rangeInCDR3.J
            )
}
